File: java_version/src/binary_tree_preorder_traversal.java
public class binary_tree_preorder_traversal {
    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> vals = new ArrayList<Integer>();
        if (root!=null){
            List<TreeNode> stack = new ArrayList<TreeNode>();
            TreeNode curr = root;
            int stack_size = -1;

            while(true){
                if (curr!=null){
                    vals.add(curr.val);
                    stack.add(curr);
                    stack_size += 1;
                    curr = curr.left;
                }
                else if(!stack.isEmpty()){
                    curr = stack.remove(stack_size);
                    stack_size -= 1;
                    curr = curr.right;
                }
                else{
                    break;
                }
            }
        }
        return vals;
    }
}

File: binary_tree_preorder_traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def preorderTraversal(self, root: TreeNode) -> List[int]:
        if not root:
            return []

        stack = []
        curr = root
        vals = []

        while True:
            if curr:
                vals.append(curr.val)
                stack.append(curr)
                curr = curr.left

            elif (stack):
                curr = stack.pop()
                curr = curr.right

            else:
                break

        return vals


